### PR TITLE
[SPARK-45409][INFRA] Pin `torch<=2.0.1`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to pin `torch` to recover Python Linter failure.

### Why are the changes needed?

`torch 2.1.0` is released and breaks `mypy` in a weird way.

- https://github.com/apache/spark/actions/runs/6399720715/job/17372994437

```
starting mypy annotations test...
annotations failed mypy checks:
/usr/local/lib/python3.9/dist-packages/torch/_dynamo/variables/tensor.py:369: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
If this issue continues with mypy master, please report a bug at https://github.com/python/mypy/issues
version: 0.982
/usr/local/lib/python3.9/dist-packages/torch/_dynamo/variables/tensor.py:369: : note: please use --show-traceback to print a traceback when reporting a bug
2
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Recover `Python Linter` in CIs.

I also manually verified this.
```
$ pip3 freeze | grep torch
torch==2.0.1
torcheval==0.0.7
torchvision==0.15.2

$ dev/lint-python
...
all lint-python tests passed!
```

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
This is `Draft` because we need to verify this.

File: dev/infra/Dockerfile [hunk 1]
```diff
@@ -91,7 +91,7 @@ RUN python3.9 -m pip install numpy pyarrow 'pandas<=2.1.1' scipy unittest-xml-re
 RUN python3.9 -m pip install 'grpcio>=1.48,<1.57' 'grpcio-status>=1.48,<1.57' 'protobuf==3.20.3' 'googleapis-common-protos==1.56.4'
 
 # Add torch as a testing dependency for TorchDistributor
-RUN python3.9 -m pip install torch torchvision --index-url https://download.pytorch.org/whl/cpu
+RUN python3.9 -m pip install 'torch<=2.0.1' torchvision --index-url https://download.pytorch.org/whl/cpu
```
Comment (HyukjinKwon) [inline dev/infra/Dockerfile:94]: LGTM if it needs. We may also better fix https://github.com/apache/spark/blob/master/dev/requirements.txt together

Comment (dongjoon-hyun) [reply dev/infra/Dockerfile:94]: Thank you. Sure.

**Review by HyukjinKwon — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you. I verified that the Python Linter passed at the second last commit.. The last commit doesn't change the CI. Let me merge this. Let me merge this to recover the master branch.


Screenshot 2023-10-03 at 11 58 41 PM

**Comment by zhengruifeng** [comment]
Late LGTM, thank you for this quick fix

**Pull request closed without merging**